### PR TITLE
Add a default argument to the datasets and a possibility to override datasets

### PR DESCRIPTION
**Pull request by nils-braun**
Following the discussion in https://github.com/dask/dask/issues/6531, that PR adds two new features:
* if the get datasets method returns nothing, a new default keyword argument on the client can prevent the KeyError (similar to `dict.get`). I also added a `get` method to the `client.datasets`.
* A new keyword argument `override` in `publish_dataset` lets you also prevent the KeyError there on duplicated entries (server side)

### CODE REVIEW

File: distributed/client.py [hunk 1]
```diff
@@ -2280,6 +2281,7 @@ def publish_dataset(self, *args, **kwargs):
         ----------
         args : list of objects to publish as name
         name : optional name of the dataset to publish
+        override : if true, override any already present dataset with the same name
```
Comment (martindurant) [inline distributed/client.py:2284]: If false, will raise when the key already exists.
These argument descriptions could use a better style, like
```
override : bool (optional, default False)
    if true, override ...
```

File: distributed/client.py [hunk 2]
```diff
@@ -2342,24 +2344,28 @@ def list_datasets(self, **kwargs):
...
         return self.sync(self.scheduler.publish_list, **kwargs)
 
-    async def _get_dataset(self, name):
+    async def _get_dataset(self, name, default):
         with self.as_current():
             out = await self.scheduler.publish_get(name=name, client=self.id)
 
         if out is None:
-            raise KeyError(f"Dataset '{name}' not found")
+            if default is None:
```
Comment (martindurant) [inline distributed/client.py:2352]: What if we want a default of `None`? Perhaps use use placeholder like `"_no_default_"`.

File: distributed/client.py [hunk 3]
```diff
@@ -2342,24 +2344,28 @@ def list_datasets(self, **kwargs):
...
+            else:
+                return default
         return out["data"]
 
-    def get_dataset(self, name, **kwargs):
+    def get_dataset(self, name, default=None, **kwargs):
         """
-        Get named dataset from the scheduler
+        Get named dataset from the scheduler if present.
+        Return the default or raise a KeyError if not present.
```
Comment (martindurant) [inline distributed/client.py:2361]: Could do with a full update to args here too

File: distributed/publish.py [hunk 4]
```diff
@@ -116,3 +116,7 @@ def __len__(self):
                 "please use 'len(await client.list_datasets())' instead"
             )
         return len(self._client.list_datasets())
+
+    def get(self, key, default=None):
```
Comment (martindurant) [inline distributed/publish.py:120]: Why is this needed?

File: distributed/tests/test_publish.py [hunk 5]
```diff
@@ -208,6 +222,18 @@ def test_datasets_contains(client):
...
+def test_datasets_republish(client):
+    key, value, value2 = "key", "value", "value2"
+    client.publish_dataset(key=value)
+    assert client.get_dataset(key) == value
+
+    with pytest.raises(KeyError) as exc_info:
+        client.publish_dataset(key=value)
+
+    client.publish_dataset(key=value2, override=True)
+    assert client.get_dataset(key) == value2
```
Comment (mrocklin) [inline distributed/tests/test_publish.py:234]: This behavior seems sensible to me

**Comment by mrocklin** [comment]
This all seems good to me.  Deferring to @martindurant for a decision on the behavior

**Comment by martindurant** [comment]
Fixes  dask/dask#6531

File: distributed/publish.py (hunk 4 above)
Comment (nils-braun) [reply distributed/publish.py:120]: Yes, you are totally correct. It will pick up the default argument anyways. Thanks for the hint!

**Comment by nils-braun** [comment]
Thanks for your review @martindurant and @mrocklin. I have implemented your feedback

File: distributed/client.py [hunk 6]
```diff
@@ -2342,24 +2347,36 @@ def list_datasets(self, **kwargs):
         """
         return self.sync(self.scheduler.publish_list, **kwargs)
 
-    async def _get_dataset(self, name):
+    async def _get_dataset(self, name, default=NO_DEFAULT_PLACEHOLDER):
```
Comment (martindurant) [inline distributed/client.py:2350]: @mrocklin , is there a utility variable or other standard way to do this?

Otherwise, happy with the PR.

Comment (nils-braun) [reply distributed/client.py:2350]: I did a quick grep through the `distributed` repo and have found

```python
_empty_namedtuple = collections.namedtuple("_empty_namedtuple", ())
```
in `metrics.py` - but I do not know if this is the standard. Happy to follow one!

Comment (mrocklin) [reply distributed/client.py:2350]: I am personally more familiar with using `__no_default__` over a named tuple.  I think that what is here is fine.

**Comment by martindurant** [comment]
(also requires running `black`, unfortunately)

**Comment by nils-braun** [comment]
Thanks @martindurant for fixing the style
